### ISSUE TITLE
Multi-camera setup: 1 head + 2 wrist — image stacking vs. future-frame conditioning

### ISSUE THREAD

**Issue by jinsonglin-cuhk** [Mondo-Robotics/DiT4DiT#12]
Hi,

In dataloader/gr00t_lerobot/datasets.py (around the block that fills prim_t0 / wrist_t0 and builds all_images), side-by-side concat only runs when len(wrist_t0) == len(prim_t0); otherwise it uses all_images = prim_t0 + wrist_t0.

With one head camera and two wrist cameras, lengths usually do not match (e.g. one primary stream vs. two wrist streams per timestep), so we do not get a single frame that spatially concatenates all three views.

Could you clarify the intended behavior for future-frame / video conditioning in this case? Is the current list fallback deliberate, or should multi-wrist setups be merged (e.g. triple-width concat or another layout) so Cosmos VAE sees one coherent multi-view frame per timestep?

Additionally, I would like to confirm: under this length-mismatch scenario, is it the case that only the first frame from the head camera is used as the current frame, with no first frame from the wrist cameras? And are the subsequent frames of the head camera plus all frames of the wrist cameras treated as future frames for prediction?

Thanks.

**Comment by TeleeMa** [comment]
Great question, thanks for the careful reading of the code.


Honestly, the fallback is not the intended design for multi-wrist setups — it was written primarily with the 1+1 layout in mind (which covers LIBERO and our RoboCasa/Real-G1 experiments). In the fallback case, all primary frames come first, followed by all wrist frames as a flat temporal sequence, which means:

- Cosmos VAE does not see a coherent multi-view frame per timestep.
- The primary and wrist streams get interpreted as a longer "single-view" temporal sequence, which breaks the intended semantics.

Your second question — which frames become "current" vs "future"?

In the 1+1 case, the concatenated list [prim⊕wrist_0, prim⊕wrist_1, ..., prim⊕wrist_{T-1}] is passed as image into Cosmos2.5. Inside the VLM forward, frame index 0 is used as the conditional (current) frame and indices ≥ 1 are treated as future frames to be denoised. This is controlled by cond_mask / cond_indicator built in prepare_latents — see [Cosmos25.py:561-576]

In the fallback (1-head + 2-wrist) case, the ordering becomes [head_0, head_1, ..., head_{T-1}, wrist1_0, ..., wrist1_{T-1}, wrist2_0, ..., wrist2_{T-1}]. Given the cond-mask convention, only the first head frame would be treated as the current frame, and everything after (remaining head frames + all wrist frames) would be treated as "future" — which is clearly not what we want. So your interpretation of the fallback behavior is correct.


A clean fix is triple-width (or K-wide) spatial concat per timestep:


```
# Group wrist views by timestep then concat spatially with the primary frame
T = len(prim_t0)
num_wrist = len(wrist_t0) // T if T > 0 else 0
if num_wrist > 0 and len(wrist_t0) == num_wrist * T:
    all_images = []
    for t in range(T):
        views = [prim_t0[t]] + [wrist_t0[t + k * T] for k in range(num_wrist)]
        all_images.append(torch.cat(views, dim=-1))  # (C, H, (1+num_wrist)*W)
else:
    all_images = prim_t0 + wrist_t0
```

This makes Cosmos VAE see one coherent multi-view frame per timestep (head | wrist_left | wrist_right side-by-side), preserving the temporal alignment.


We'll fix this till next code commit, thanks for flagging this clearly!

**Comment by TeleeMa** [comment]
@jinsonglin-cuhk fixed now.
